presidio-recognizers-store/cmd/presidio-recognizers-store/main.go

```diff
@@ -0,0 +1,348 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"strconv"
+	"time"
+
+	"google.golang.org/grpc/reflection"
+
+	"flag"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	types "github.com/Microsoft/presidio-genproto/golang"
+
+	"github.com/Microsoft/presidio/pkg/cache"
+	log "github.com/Microsoft/presidio/pkg/logger"
+	"github.com/Microsoft/presidio/pkg/platform"
+	"github.com/Microsoft/presidio/pkg/presidio/services"
+	"github.com/Microsoft/presidio/pkg/rpc"
+)
+
+type server struct{}
+
+var (
+	settings         *platform.Settings
+	recognizersStore cache.Cache
+)
+
+var (
+	recognizersKey = "custom_recognizers"
+	timestampKey   = "custom_recognizers:last_update"
+)
+
+func main() {
+
+	pflag.Int(platform.GrpcPort, 3004, "GRPC listen port")
+	pflag.String(platform.RedisURL, "localhost:6379", "Redis address")
+	pflag.String(platform.RedisPassword, "", "Redis db password (optional)")
+	pflag.Int(platform.RedisDb, 0, "Redis db (optional)")
+	pflag.Bool(platform.RedisSSL, false, "Redis ssl (optional)")
+	pflag.String(platform.PresidioNamespace, "", "Presidio Kubernetes namespace (optional)")
+	pflag.String("log_level", "info", "Log level - debug/info/warn/error")
+
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+	pflag.Parse()
+	viper.BindPFlags(pflag.CommandLine)
+
+	settings = platform.GetSettings()
+	log.CreateLogger(settings.LogLevel)
+
+	svc := services.New(settings)
+
+	if settings.RedisURL != "" {
+		recognizersStore = svc.SetupCache()
+	}
+
+	lis, s := rpc.SetupClient(settings.GrpcPort)
+
+	types.RegisterRecognizersStoreServiceServer(s, &server{})
+	reflection.Register(s)
+
+	if err := s.Serve(lis); err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+// setValue sets the recognizers array in the store and updates the last update
+// timestamp
+func setValue(value string, timestamp string) error {
+	err := recognizersStore.Set(recognizersKey, value)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	log.Info("Updating timestamp: " + timestamp)
+	err = recognizersStore.Set(timestampKey, timestamp)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func getExistingRecognizers() ([]types.PatternRecognizer, error) {
+	recognizersArr := make([]types.PatternRecognizer, 0)
+	existingItems, err := recognizersStore.Get(recognizersKey)
+	if err == nil && existingItems != "" {
+		log.Info("Found existing recognizers...")
+		var recognizers []types.PatternRecognizer
+		err = json.Unmarshal([]byte(existingItems), &recognizers)
+		if err != nil {
+			return nil, err
```

Comment:
If the key contains information that cannot be unmarshalled the system will remain in a failed state until there is some user interaction to clear down the key. 

Maybe one option would be to clear down the key data if it's in a bad state before returning the error.

Reply:
A valid point. however, this can happen for example due to a change in the types.PatternRecognizer class which is not backward compatible or any other issue.
Deleting the recognizer (from the storage) as a self healing solution can be done, however this can lead to lose of configuration which is irreversible.

I think that in the current suggested implementation, if the unmarshaling fails you get an error logged and the process continues without the custom recognizers, so the system remains fully operational, until some user manually fix it, or a new version is deployed (if thats the case)

@omri374  WDYT?

Reply:
If there's information that cannot be unmarshalled, we'd lose it anyway, wouldn't we? Services would not be able to parse it and return it. However I agree that the system can still remain functional (although it's not ideal). @eladiw let's discuss this offline and decide what would be the best way.

Reply:
Discussed. Agreed that 'auto recovery' is not feasible and should not be a priority at this point.
Also, deleting these items may pose bigger risk than just ignoring.

Decided to leave as is